File: problems/0023-merge-k-sorted-lists/typescript/solution.ts
class ListNode {
  val: number;
  next: ListNode | null = null;
  constructor(val = 0, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function build(vals: number[]): ListNode | null {
  const dummy = new ListNode();
  let cur = dummy;
  for (const v of vals) {
    cur.next = new ListNode(v);
    cur = cur.next;
  }
  return dummy.next;
}

function toArray(head: ListNode | null): number[] {
  const out: number[] = [];
  for (let n = head; n !== null; n = n.next) out.push(n.val);
  return out;
}

function mergeTwo(l1: ListNode | null, l2: ListNode | null): ListNode | null {
  const dummy = new ListNode();
  let tail = dummy;
  while (l1 !== null && l2 !== null) {
    if (l1.val <= l2.val) {
      tail.next = l1;
      l1 = l1.next;
    } else {
      tail.next = l2;
      l2 = l2.next;
    }
    tail = tail.next;
  }
  tail.next = l1 !== null ? l1 : l2;
  return dummy.next;
}

// Merge k Sorted Lists — https://leetcode.com/problems/merge-k-sorted-lists/
function mergeKLists(lists: (ListNode | null)[]): ListNode | null {
  if (lists.length === 0) return null;
  while (lists.length > 1) {
    const merged: (ListNode | null)[] = [];
    for (let i = 0; i < lists.length; i += 2) {
      const l2 = i + 1 < lists.length ? lists[i + 1] : null;
      merged.push(mergeTwo(lists[i], l2));
    }
    lists = merged;
  }
  return lists[0];
}

export function mergeKListsArr(arrays: number[][]): number[] {
  return toArray(mergeKLists(arrays.map((a) => build(a))));
}

File: problems/0023-merge-k-sorted-lists/python/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def build(vals):
    dummy = ListNode()
    cur = dummy
    for v in vals:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next


def to_list(head):
    out = []
    while head:
        out.append(head.val)
        head = head.next
    return out


def mergeTwo(l1, l2):
    dummy = ListNode()
    tail = dummy
    while l1 and l2:
        if l1.val <= l2.val:
            tail.next = l1
            l1 = l1.next
        else:
            tail.next = l2
            l2 = l2.next
        tail = tail.next
    tail.next = l1 if l1 else l2
    return dummy.next


def mergeKLists(lists):
    """Merge k Sorted Lists — https://leetcode.com/problems/merge-k-sorted-lists/"""
    if not lists:
        return None
    while len(lists) > 1:
        merged = []
        for i in range(0, len(lists), 2):
            l2 = lists[i + 1] if i + 1 < len(lists) else None
            merged.append(mergeTwo(lists[i], l2))
        lists = merged
    return lists[0]


def mergeKListsArr(arrays):
    return to_list(mergeKLists([build(a) for a in arrays]))
